flixel/graphics/FlxAsepriteUtil.hx

```diff
@@ -0,0 +1,335 @@
+package flixel.graphics;
+
+import flixel.graphics.frames.FlxAtlasFrames;
+import flixel.system.FlxAssets;
+
+/**
+ * Collection of helpers that deal with Aseprite files. Namely the json files exported for sprite sheets.
+ */
+class FlxAsepriteUtil
+{
+	/**
+	 * Helper for parsing Aseprite atlas json files. Reads frames via `FlxAtlasFrames.fromAseprite`
+	 * and returns the parsed AseAtlas to be used with other `FlxAsepriteUtil` helpers.
+	 * 
+	 * @param   sprite   The sprite to load the ase atlas's frames
+	 * @param   graphic  The png file associated with the atlas
+	 * @param   data     Can be an `AseAtlas` struct, a JSON string matching the `AseAtlas` or a string asset path to a json
+	 * @return  This `FlxSprite` instance (nice for chaining stuff together, if you're into that).
+	 * @see flixel.graphics.FlxAsepriteUtil.AseAtlasMeta
+	 * @since 5.4.0
+	 */
+	public static function loadAseAtlas(sprite:FlxSprite, graphic, data:FlxAsepriteJsonAsset)
+	{
+		sprite.frames = FlxAtlasFrames.fromAseprite(graphic, data);
+		return sprite;
+	}
+	
+	/**
+	 * Helper for parsing Aseprite atlas json files. Reads frame data via `FlxAtlasFrames.fromAseprite`,
+	 * then, adds animations for any tags listed.
+	 * 
+	 * @param   sprite     The sprite to load the ase atlas's frames
+	 * @param   graphic    The png file associated with the atlas
+	 * @param   data       Can be an `AseAtlas` struct, a JSON string matching the `AseAtlas` or a string asset path to a json
+	 * @param   tagSuffix  The delimeter on each frame name that follows the animation name and precedes the frame number
+	 * @return  This `FlxSprite` instance (nice for chaining stuff together, if you're into that).
+	 * @see flixel.graphics.FlxAsepriteUtil.AseAtlasMeta
+	 * @since 5.4.0
+	 */
+	public static function loadAseAtlasAndTags(sprite:FlxSprite, graphic, data:FlxAsepriteJsonAsset, tagSuffix:String = ":")
+	{
+		final aseData = data.getData();
+		loadAseAtlas(sprite, graphic, aseData);
+		return addAseAtlasTags(sprite, aseData, tagSuffix);
+	}
+	
+	/**
+	 * Loops through the given ase atlas's tags and adds animations for each, to the gives sprite.
+	 * 
+	 * @param   sprite     The sprite to add the animations
+	 * @param   data       Can be an `AseAtlas` struct, a JSON string matching the `AseAtlas` or a string asset path to a json.
+	 * @param   tagSuffix  The delimeter on each frame name that follows the animation name and precedes the frame number
+	 * @return  This `FlxSprite` instance (nice for chaining stuff together, if you're into that).
+	 * @see flixel.graphics.FlxAsepriteUtil.AseAtlasMeta
+	 * @since 5.4.0
+	 */
+	public static function addAseAtlasTags(sprite:FlxSprite, data:FlxAsepriteJsonAsset, tagSuffix:String = ":")
+	{
+		final aseData = data.getData();
+		for (frameTag in aseData.meta.frameTags)
+			sprite.animation.addByPrefix(frameTag.name, frameTag.name + tagSuffix);
```

Comment:
A question here as this was part of the work I had to do to get this working properly. `addByPrefix` relies on the existing loaded frame names. The `frameTag.name` is the name given to the animation tag in Aseprite (such as `run_left`, we'll say for this example).

The frame names (at least with the fairly default export options I'm using), are loaded based on the file name that they came from. My example is that I have a `player.ase` file, and the export json has the frames named like `player 0.ase`, `player 1.ase`, etc. So if you do `addByPrefix("run_left", "run_left:")`, there are no frames that match.

In my case, I had to build the animations based on the frame index data on the tag (`from` and `to`): https://github.com/bitDecayGames/haxeflixel-template/blob/master/source/loaders/Aseprite.hx#L54-L57

I haven't handled the case where you export with the default `--json-hash` option explicitly. There may be some strangeness with the order frames are added depending on how the hash is initially iterated when adding the frames.

Am I off-base with this, or does it make sense that this function won't properly build out animations as it sits?

Reply:
<img width="439" alt="Screenshot 2023-07-18 at 1 09 04 PM" src="https://github.com/HaxeFlixel/flixel/assets/2609513/b5a73306-5b2f-47c8-884a-ce22645aa2e3">

I'm not familiar with the batch exporter, but in the editor you can specify a format for the frame names, this is the format to use with the default delimiter. I was gonna add this to the doc

> In my case, I had to build the animations based on the frame index data on the tag (from and to):

that's not a bad way to go about it either, assuming the frames are always ordered, predictably.

Reply:
Ah, I definitely missed that as I'm only using the CLI at the moment. They have `--tag-format` documented, but don't explicitly call out that it follows the same formatting rules that `--file-format` uses. However, if you click the `(?)` in the UI, it takes you here: https://www.aseprite.org/docs/cli/#filename-format

One thing I noticed with your export settings (which I think actually makes the frames way more readable than the default of `{title} {frame}.{extension}`) is that when you have overlapping tags, you lose information, for example:

I have two animations `a` and `b`, and I am using one frame of animation `a` as my knockback frame when my character takes damage
![image](https://github.com/HaxeFlixel/flixel/assets/3443665/c791943d-7590-4b9b-8cd5-5a3e1ffc980f)

This exports as (shorthanded):
```
  "frames": {
   "a:0": {...},
   "b:0": {...},
   "a:2": {...},
   .
   .
   .
   },
 },
 ```
So you would lose that frame from your `a` animation if you simply load them by prefix.

I think using the `from` and `to` are the proper way to do this. They are easy to handle if you do `Array` export (via the drop-down in the UI, or via the `--json-array` flag on CLI.

The tricky part is if you select `Hash` in the UI (which is also the default for the CLI. It looks like you (or whoever originally implemented it if you pulled it from somewhere else) has the `abstract Hash<T>(Dynamic)` type that gets a key value iterator via `Reflect.fields(this).iterator();`. I'm still looking if that has any guarantees of field ordering (seems a little dangerous, either way due to potential platform differences). I'm curious why Hash is the default since there is an inherently linear nature to animation frames. 🤔 Maybe because you can have empty frames that don't have to be exported, so you can avoid sparsely populated arrays?

Reply:
Well I'm finding more weird edge-cases. If you add an empty frame to your animation (frame 5 on this timeline):
![image](https://github.com/HaxeFlixel/flixel/assets/3443665/b95799bd-fd15-4bf8-bf3d-0b0065578436)

Then export the sprite sheet and check the `Layout` -> `Ignore Empty` box

Your frames are exported as:
```
  "frames": {
   "a:0": {...},
   "a:1": {...},
   "a:2": {...},
   "a:3": {...},
   "a:5": {...},
   "left:0": {...},
   "left:1": {...},
   "left:2": {...},
   "left:3": {...}
 },
```

And your `frameTag` for `a` looks like:
```
{ "name": "a", "from": 0, "to": 5, ...}
```

So there is an implicit frame index referenced in the tag that isn't actually exported in the json. The two tags `a` and `left` reference 10 total frames across their `from` and `to` values, but there are only 9 elements in the `frames` Hash/Array.

Maybe it's best to just document what export settings are supported and draw some reasonable bounds to keep the scope reasonable.

Reply:
> To be honest, this seems like an Aseprite bug to me.

<img width="288" alt="Screenshot 2023-07-18 at 2 13 17 PM" src="https://github.com/HaxeFlixel/flixel/assets/2609513/1d84f04a-11a9-4afe-9806-bb5ba335b1d0">

<img width="575" alt="Screenshot 2023-07-18 at 2 13 54 PM" src="https://github.com/HaxeFlixel/flixel/assets/2609513/31dfaec8-9177-4a35-b527-669c52909549">

There is an option to ignore empty frames

Reply:
> 

I was calling out specifically when there is an empty frame in the middle of a tag, it will _not include_ the frame (as expected if you check that box), but it will give your frame tag counts as though that frame still exists. I think it's well beyond the scope of what you're supporting with this PR. Here's a simplified example just to exhibit the weirdness without any additional stuff:

Simple animation with 3 frames. Middle frame is empty:
![image](https://github.com/HaxeFlixel/flixel/assets/3443665/998992a9-1a50-49f2-8855-083f0e597210)

Export with `Ignore Empty` checked:
![image](https://github.com/HaxeFlixel/flixel/assets/3443665/db409e40-1be3-41ab-bc10-ee243fad79a8)

Shorthanded JSON:
```
{
  "frames": [
   {
    "filename": "a:0",...},
   {
    "filename": "a:2",...}
 ],
 "meta": {
  "frameTags": [
   { "name": "a", "from": 0, "to": 2, ...}
  ],
 }
}
```

Tag `a` says it is from frame `0` to frame `2` (these are frame indexes rather than names), which is 3 frames total. However, there are only 2 frames in the frame array. So our tag references a frame that aseprite took out during export, but didn't adjust the tag metadata, which _should_ be `from: 0, to: 1`.

But I'm way too deep in edge cases at the moment. I'm bubbling back up to just focus on the high level best way to parse tag frames out as animations.

At the moment, your way requires that you set the `Item Filename` to something other than the default for parsing to work (your `{tag}:{tagframe}` format). This will miss frames if the user has overlapping tags in their .ase file.

The other option is to go off of the `from` and `to` in the `frameTag` data, but that has the _potential_ issue of how fields are parsed when `Hash` is used.

Do you have a preference or general thoughts on those two directions?

Reply:
also, there's "{outertag}:{frame}" which always uses the longest tag which solves your issue above with `"a:0", "b:0" "a:2"` outertag would result in `"a:0", "a:1" "a:2"`

Reply:
Yeah, it seems like either way, you lose information needed to parse this easily. Using `{tag}`, you lose the ability to parseByPrefix to build tag `a` properly. Using `{outertag}` loses that tag `b` even exists. 🤔 This is way more subtle than I was expecting.

And yes, the quick summary of my above rant is that ignoring empty frames makes it so the frame indexes in the tag are unreliable. Which seems really silly - it should be reasonably simple to account for having removed empty frames so those indices are correct.